internal/wal_segment.go

```diff
@@ -0,0 +1,222 @@
+	endLSN          LSN
+	walSegmentBytes uint64
+	data            []byte
+	readIndex       int
+	writeIndex      int
+	lastMsg         *pgproto3.BackendMessage
+}
+
+// The ProcessMessageResult is an enum representing possible results from the methods processing the messages as received from Postgres into the wal segment.
+type ProcessMessageResult int
+
+// These are the multiple results that the methods can return
+const (
+	ProcessMessageOK ProcessMessageResult = iota
+	ProcessMessageUnknown
+	ProcessMessageCopyDone
+	ProcessMessageReplyRequested
+	ProcessMessageSegmentGap
+	ProcessMessageMismatch
+)
+
+// NewWalSegment is a helper function to declare a new WalSegment.
+func NewWalSegment(timeline int32, location LSN, walSegmentBytes uint64) *WalSegment {
+	//We could test validity of walSegmentBytes (not implemented):
+	//  https://www.postgresql.org/docs/11/app-initdb.html:
+	//    Set the WAL segment size, in megabytes...
+	//    The value must be a power of 2 between 1 and 1024 (megabytes)
+
+	segment := &WalSegment{TimeLine: timeline, walSegmentBytes: walSegmentBytes}
+	// Calculate start byte of file from location (which could be anywhere in this file)
+	segment.StartLSN = LSN((uint64(location) / walSegmentBytes) * walSegmentBytes)
+	// Calculate end form start and number of bytes in this file
+	segment.endLSN = segment.StartLSN + LSN(walSegmentBytes)
+	// Allocate data
+	segment.data = make([]byte, walSegmentBytes)
+	return segment
+}
+
+// NextWalSegment is a helper function to create the next wal segment which comes after this wal segment.
+// Note that this will be on the same timeline. the convenience is that it also automatically processes
+// a message that crosses the boundary between the two segments.
+func (seg *WalSegment) NextWalSegment() (*WalSegment, error) {
+	// Next on this timeline, but read rest of msg
+	if ! seg.IsComplete() {
+		return nil, segmentError{
+			errors.Errorf("Cannot run NextWalSegment until IsComplete")}
+	}
+	nextSegment := NewWalSegment(seg.TimeLine, seg.endLSN, seg.walSegmentBytes)
+	if seg.lastMsg != nil {
+		// Apparaently the last message crossed the border between the two segments, so lets have it processed into the next segment too.
+		result, err := nextSegment.ProcessMessage(*seg.lastMsg)
+		if err != nil {
+			return nil, err
+		}
+		if result != ProcessMessageOK {
+			return nil, segmentError{
+				errors.Errorf("Unexpected result from ProcessMessage in NextWalSegment")}
+		}
+	}
+	return nextSegment, nil
+}
+
+// Name returns the filename of this wal segment.
+// This is also used by the WalUploader to set the name of the destination file during upload of the wal segment.
+func (seg *WalSegment) Name() string {
+	// Example LSN -> Name:
+	// '0/2A33FE00' -> '00000001000000000000002A'
+	segID := uint64(seg.StartLSN)/uint64(seg.walSegmentBytes)
+	if seg.IsComplete() {
+		return fmt.Sprintf("%08X%016X", seg.TimeLine, segID)
+	}
+	return fmt.Sprintf("%08X%016X.partial", seg.TimeLine, segID)
```

Comment:
Recieveing timeline switch record ia a bit tricky.
Keep in mind that we need to archive partial segment on old timeline and full segmet on new timeline.

Reply:
Good thinking. I am now checking, and I don't see the partial segments being written out. Will fix.

Reply:
Fixed.